students/guillaume/quarter_2_project/leankitbackend/leankitbackend/dataprocess.py

```diff
@@ -0,0 +1,311 @@
+        lst = [sprint[0]]
+        lst.extend(ret_lst)
+        data.append(lst)
+    table = create_table(data, lanes_lst)
+    return table
+
+
+def lst_projects(Cards_obj):
+
+    projects_lst = projects(Cards_obj)
+    cards_proj = cards_per_projects(Cards_obj)
+
+    data = []
+    for project in projects_lst:
+        i = project_size(cards_proj[project])
+        j = project_in_work(cards_proj[project])
+        lst = [project.title, project.plannedFinish,
+               project.lane['title'], i, j, assigned(project)]
+        data.append(lst)
+    data.sort(key=lambda x: x[2])
+    table = create_table(data, ['Project', 'End Date', 'Lane',
+                                'Total Points', 'In Work', 'Assigned to'])
+    return table
+
+
+def projects_contributors(Cards_obj):
+
+    projects = cards_per_projects(Cards_obj)
+
+    def user(cards_lst):
+        ret_lst = []
+        if not isinstance(cards_lst, list):
+            cards_lst = [cards_lst]
+        for card in cards_lst:
+            for assigne in card.assignedUsers:
+                ret_lst.append(assigne['fullName'])
+        return ret_lst
+
+    ret_lst = []
+    for project, cards_lst in projects.items():
+        i = project_size(cards_lst)
+        j = project_in_work(cards_lst)
+        list_users = user(project)
+        if cards_lst != []:
+            list_users.extend(user(cards_lst))
+        # using set for unicity purpose
+        tmp = [project.title, project.lane['title'], project.plannedFinish,
+               assigned(project), len(cards_lst), i, j]
+        tmp.extend(set(list_users))
+        ret_lst.append(tmp)
+
+    nb_lane = max([len(item) for item in ret_lst])
+    fields = ['Project', 'status', 'Finish Date', 'Assigned to', '# Cards',
+              'Total Points', 'In Work']
+    fields_tmp = ['contributor_{}'.format(k + 1)
+                  for k in range(nb_lane - len(fields))]
+    fields.extend(fields_tmp)
+
+    ret_lst.sort(key=lambda x: x[1])
+
+    def ext(line):
+        line.extend([' '] * (nb_lane - len(line)))
+        return line
+
+    data = [ext(line) for line in ret_lst]
+    table = create_table(data, fields)
+    ret_lst.insert(0, fields)
+
+    return ret_lst, table
+
+
+def table_new_requests(cards, cutoff_date):
+    cards = new_requests(cards, cutoff_date)
+
+    def lst_card(card):
+        return [card.type['title'], card.title, created_by(card),
+                assigned(card), created_on(card)]
+
+    print_lst = [lst_card(card) for card in cards]
+    print_lst.sort(key=lambda x: x[-1])
+
+    header = ['Request Type', 'Title', 'Created By', 'Assigned to',
+              'Created on']
+
+    return create_table(print_lst, header)
+
+
+def create_table(iter, header=None):
+    table = PrettyTable()
+    table.field_names = header
+    for lst in iter:
+        table.add_row(lst)
+    return table
+
+
+def table_cards_per_users(users_cards, users_lst=None):
+
+    def c_data(card):
```

Comment:
same as above -- I don't think this needs to be a closure, and it's not very efficient to re-define the function each time it's used.